File: fraud-detection/src/main/java/com/wslimi/demo/fraud/processor/NaiveFraudDetectionProcessor.java
package com.wslimi.demo.fraud.processor;

import com.wslimi.demo.fraud.model.FraudNaiveAlert;
import com.wslimi.demo.fraud.model.Transaction;
import lombok.extern.slf4j.Slf4j;
import org.apache.flink.streaming.api.functions.KeyedProcessFunction;
import org.apache.flink.util.Collector;

/**
 * Simple fraud detection processor.
 * Flags a transaction as fraudulent if the amount exceeds a threshold.
 */
@Slf4j
public class NaiveFraudDetectionProcessor extends KeyedProcessFunction<String, Transaction, FraudNaiveAlert> {

    private static final double FRAUD_THRESHOLD = 10000.0;

    @Override
    public void processElement(Transaction transaction,
                               KeyedProcessFunction<String, Transaction, FraudNaiveAlert>.Context context,
                               Collector<FraudNaiveAlert> collector) throws Exception {

        if (transaction.amount() > FRAUD_THRESHOLD) {
            log.warn("🚨 FRAUD DETECTED! Transaction {} from account {} with amount {} exceeds threshold {}",
                    transaction.transactionId(),
                    transaction.srcAccountId(),
                    transaction.amount(),
                    FRAUD_THRESHOLD);

            // Emit fraudulent transaction
            FraudNaiveAlert alert = new FraudNaiveAlert(transaction.transactionId(), System.currentTimeMillis(), transaction, "Amount exceeds threshold");
            collector.collect(alert);
        } else {
            log.info("✅ Transaction {} is valid. Amount: {}",
                    transaction.transactionId(),
                    transaction.amount());
        }
    }
}

File: fraud-detection/src/main/python/processor/naive_fraud_detection_processor.py
"""
Naive Fraud Detection Processor - Simple threshold-based fraud detection.
"""
import logging
import uuid

from pyflink.datastream import KeyedProcessFunction

from model.transaction import Transaction
from model.fraud_naive_alert import FraudNaiveAlert


logger = logging.getLogger(__name__)


class NaiveFraudDetectionProcessor(KeyedProcessFunction):
    """
    Simple fraud detection processor.
    Flags a transaction as fraudulent if the amount exceeds a threshold.
    This is a STATELESS approach - no pattern detection, just a simple threshold check.
    """

    FRAUD_THRESHOLD = 10000.0

    def process_element(self, transaction: Transaction, context: 'KeyedProcessFunction.Context'):
        """
        Process each transaction and check if it exceeds the fraud threshold.

        Args:
            transaction: The transaction to process
            context: Flink process context
        """
        if transaction.amount > self.FRAUD_THRESHOLD:
            logger.warning(
                "🚨 FRAUD DETECTED! Transaction %s from account %s with amount %s exceeds threshold %s",
                transaction.transaction_id,
                transaction.src_account_id,
                transaction.amount,
                self.FRAUD_THRESHOLD
            )

            # Emit fraudulent transaction
            alert = FraudNaiveAlert(
                alert_id=str(uuid.uuid4()),
                timestamp=context.timer_service().current_processing_time(),
                transaction=transaction,
                comment=f"Amount exceeds threshold of {self.FRAUD_THRESHOLD}"
            )
            yield alert
        else:
            logger.info(
                "✅ Transaction %s is valid. Amount: %s",
                transaction.transaction_id,
                transaction.amount
            )

